libafl/src/executors/forkserver.rs

```diff
@@ -0,0 +1,452 @@
+                rlim_cur: memlimit,
+                rlim_max: memlimit,
+            };
+            let r0 = libc::rlimit {
+                rlim_cur: 0,
+                rlim_max: 0,
+            };
+
+            let ret = unsafe { libc::setrlimit(libc::RLIMIT_AS, &r) };
+            if ret < 0 {
+                panic!("setrlimit() failed");
+            }
+            let ret = unsafe { libc::setrlimit(libc::RLIMIT_CORE, &r0) };
+            if ret < 0 {
+                panic!("setrlimit() failed");
+            }
+            Ok(())
+        };
+        unsafe { self.pre_exec(func) }
+    }
+}
+
+pub struct OutFile {
+    file: File,
+}
+
+impl OutFile {
+    pub fn new(file_name: &str) -> Self {
+        let f = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(file_name)
+            .expect("Failed to open the input file");
+        Self { file: f }
+    }
+    pub fn as_raw_fd(&self) -> RawFd {
+        self.file.as_raw_fd()
+    }
+
+    pub fn write_buf(&mut self, buf: &Vec<u8>) {
+        self.file.seek(SeekFrom::Start(0)).unwrap();
+        self.file.write(buf).unwrap();
+        self.file.set_len(buf.len() as u64).unwrap();
+        self.file.flush().unwrap();
+    }
+
+    pub fn rewind(&mut self) {
+        self.file.seek(SeekFrom::Start(0)).unwrap();
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Pipe {
+    read_end: RawFd,
+    write_end: RawFd,
+}
+
+impl Pipe {
+    fn new() -> Self {
+        let mut fds = [-1 as c_int, -1 as c_int];
+        let ret = unsafe { libc::pipe(fds.as_mut_ptr()) };
+        if ret < 0 {
+            panic!("pipe() failed");
+        }
+        Self {
+            read_end: fds[0],
+            write_end: fds[1],
+        }
+    }
+}
+
+impl Drop for Pipe{
+    fn drop(&mut self){
+        unsafe {
+            libc::close(self.read_end);
+            libc::close(self.write_end);
+            libc::close(self.read_end);
+            libc::close(self.write_end);
+        }
+    }
+}
+
+
+pub struct Forkserver {
+    st_pipe: Arc<Pipe>,
+    ctl_pipe: Arc<Pipe>,
+    pid: u32, //forkserver pid
+    child_pid: u32, //fuzzed program pid
+    status: i32,
+    last_run_timed_out: i32,
+}
+
+
+impl Forkserver {
+    pub fn new(target: String, args: Vec<String>, out_filefd: RawFd, use_stdin: bool, memlimit: u64) -> Self {
+        //check if we'll use stdin
+
+
+        let mut pid = 0;
+        //create 2 pipes
+        let st_pipe = Arc::new(Pipe::new());
+        let ctl_pipe = Arc::new(Pipe::new());
+
+        //setsid, setrlimit, direct stdin, set pipe, and finally fork.
+        match Command::new(target)
+            .args(args)
+            .stdin(Stdio::null())
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .env("LD_BIND_LAZY", "1")
+            .setlimit(memlimit)
+            .setsid()
+            .setstdin(out_filefd, use_stdin)
+            .setpipe(st_pipe.clone(), ctl_pipe.clone())
```

Comment:
I think we don't need `Arc` here, read and write ends are independent of each other

Reply:
but if we don't wrap it in Arc then Pipe's drop() would be prematurely called after `setpipe`, and following communication through the ctl_pipe and st_pipe won't work.

Reply:
ah, ok.
I guess it would simpler to pass four RawFds (st_pipe.read_end, st_pipe.write_end, ctl_pipe.read_end, ctl_pipe.write_end) to setpipe, then no Arc is needed

Reply:
Wouldn't it on drop the pipe in the parent after the child got spawned, in which case it is fine, though? Or does it drop before exec, too?
Maybe we need to `ManuallyDrop`, then?
In any case we shouldn't need `Arc` since we don't have more than 1 access to each side of the pipe I think (worst case, `Rc`?)

Reply:
it drops after the forkserver is spawned (so after the match Command::new(target) line) for the parent,
```
        unsafe {
            libc::close(ctl_pipe.read_end);
            libc::close(st_pipe.write_end);
        }
```
these 2 are safe to close for the parent, but if the Pipe::drop() is called here, the remaining two ends (ctl_pipe.write_end, st_pipe.read_end) would also be closed, and we don't want that.
what we want instead is calling Pipe::drop() when the forkserver dies.
I'll look into `ManuallyDrop`.
I tried to use `Rc`, but `pre_exec` in `std::os::unix::process::CommandExt` wants `std::marker::Send`, so it didn't work.

Reply:
We can change the `Pipe` interface to anything that makes sense.
I feel like we should move out the two ends for the forkserver, so they get dropped in the parent, and keep the respective other ends around in the fs parent, right?